contracts/dual_token/DualStakingToken.sol

```diff
@@ -61,33 +64,47 @@ contract DualStakingToken {
     // Owner address
     address public owner;
 
+    // Reentrancy lock
+    uint256 internal _locked = 1;
+
     // Mapping of staker account address => Staker info struct
-    mapping(uint256 => StakerInfo) public mapStakerStakerInfos;
+    mapping(uint256 => StakerInfo) public mapStakerInfos;
     /// Mapping of staked OLAS service multisigs
     mapping(address => bool) public mapMutisigs;
 
     /// @dev DualStakingToken constructor.
     /// @param _serviceRegistry Service registry address.
     /// @param _stakingToken Staking token address.
-    /// @param _stakingTokenAmount Staking token amount.
+    /// @param _stakingInstance Service staking instance address.
     /// @param _rewardRatio Staking token ratio to OLAS rewards in 1e18 form.
-    constructor(address _serviceRegistry, address _stakingToken, uint256 _stakingTokenAmount, uint256 _rewardRatio) {
+    constructor(
+        address _serviceRegistry,
+        address _stakingToken,
+        address _stakingInstance,
+        uint256 _rewardRatio
+    ) {
         // Check for zero addresses
-        if (_serviceRegistry == address(0) || _stakingToken == address(0)) {
+        if (_serviceRegistry == address(0) || _stakingToken == address(0) || _stakingInstance == address(0)) {
             revert ZeroAddress();
         }
 
         // Check for zero values
-        if (_stakingTokenAmount == 0 || _rewardRatio == 0) {
+        if (_rewardRatio == 0) {
             revert ZeroValue();
         }
 
         serviceRegistry = _serviceRegistry;
         stakingToken = _stakingToken;
+        stakingInstance = _stakingInstance;
 
-        stakingTokenAmount = _stakingTokenAmount;
         rewardRatio = _rewardRatio;
 
+        // Calculate staking token amount based on staking instance service information
+        uint256 numAgentInstances = IStaking(_stakingInstance).numAgentInstances();
+        uint256 minStakingDeposit = IStaking(_stakingInstance).minStakingDeposit();
+        // Total service deposit = minStakingDeposit + minStakingDeposit * numAgentInstances
+        stakingTokenAmount = (minStakingDeposit * (1 + numAgentInstances) * _rewardRatio) / 1e18;
```

Comment:
The default staking token amount is proportional to the OLAS staking amount. It can be set to a different value in the setter function.